slides/advanced-02-racing.qmd

```diff
@@ -0,0 +1,349 @@
+
+An error estimate is measured for each of 10 resamples. 
+
+ - The lines connect resamples. 
+
+There is usually a significant resample-to-resample effect (rank corr: `r round(race_cor, 2)`). 
+
+:::
+
+::: {.column width="50%"}
+
+
+```{r}
+#| label: race-data
+#| echo: false
+#| out-width: 100%
+#| fig-align: center
+#| fig-width: 5
+#| fig-height: 5
+
+
+race_example %>%
+  ggplot(aes(x = model, y = error, group = resample, col = factor(resample))) +
+  geom_point(show.legend = FALSE) +
+  geom_line(alpha = 1 / 2, show.legend = FALSE) +
+  labs(x = NULL, y = "model error")
+```
+
+:::
+
+::::
+
+
+## Are Candidates Different?
+
+One way to evaluate these models is to do a paired t-test
+ 
+ - or a t-test on their differences matched by resamples
+
+With $n = 10$ resamples, the confidence interval is (`r signif(race_ci$lower[9], 2)`, `r signif(race_ci$upper[9], 2)`), indicating that candidate number 2 has smaller error. 
+
+What if we were to compare each model candidate to the current best at each resample? 
+
+One shows superiority when `r which.max(race_ci$lower > 0)` resamples have been evaluated.
+
+
+## Evaluating Differences in Candidates
+
+```{r}
+#| label: race-ci
+#| echo: false
+#| out-width: 70%
+#| fig-align: center
+#| fig-width: 6
+#| fig-height: 4.25
+
+
+race_ci %>% 
+  ggplot(aes(`number of resamples`, difference)) + 
+  geom_point() +
+  geom_hline(yintercept = 0, col = "red", lty = 2) +
+  geom_errorbar(aes(ymin = lower, ymax = upper), width = 1 / 10, alpha = 0.5) +
+  labs(y = "error difference from best")
+```
+
+## Interim Analysis of Results
+
+One version of racing uses a _mixed model ANOVA_ to construct one-sided confidence intervals for each candidate versus the current best. 
+
+Any candidates whose bound does not include zero are discarded. 
+
+The resamples are analyzed in a random order. 
+
+[Kuhn (2014)](https://arxiv.org/abs/1405.6974) has examples and simulations to show that the method works. 
+
+The finetune package has functions `tune_race_anova()` and `tune_race_win_loss()`. 
+
+
+## Racing
+
+
+```{r tuning} 
+#| label: lgb-grid-race
+#| cache: true
+library(finetune)
+
+reg_metrics <- metric_set(mae, rsq)
+
+set.seed(9)
+lgbm_race_res <-
+  lgbm_wflow %>%
+  tune_race_anova(
+    resamples = hotel_rs,
+    grid = 25,
+    metrics = reg_metrics
+  )
+```
+
+The syntax and helper functions are extremely similar to those shown for `tune_grid()`. 
+
+
+## Racing Results
+
+:::: {.columns}
+
+::: {.column width="50%"}
+Only `r sum(map_int(lgbm_race_res$.metrics, ~ nrow(.x) / 2))` models were fit (out of 250). 
+
+`select_best()` never considers candidate models that did not get to the end of the race. 
+
+There is a helper function to see how candidate models were removed from consideration. 
+
+:::
+
+::: {.column width="50%"}
+
+
+```{r}
+#| label: plot-race
+#| out-width: 100%
+#| fig-align: center
+#| fig-width: 5
+#| fig-height: 3.8
+
+plot_race(lgbm_race_res)
```

Comment:
Is the "Analysis Stage" the number of resamples evaluated before the model was no longer evaluated? If so, ideally x axis breaks would be integers.

Reply:
I agree. I'll put an issue in to finetune for that. I added a `scale_x_continuous(breaks= pretty_breaks())`